### DISCUSSION TITLE
Specify MorphTo model

### DISCUSSION THREAD

**Discussion by GautierDele** [laravel/framework#52336] [Ideas]
Hello there,

In my use of Laravel I have several problem coming from the `morphTo` relation because i'm not able to know what's related (this is the only relation behaving like this).
A solution would be to use a `belongsTo` relation but this breaks the SQL constraints of having a single model related
In my opinion it would be great to be able to specify optionnally what model you want from the relation so you avoid misleading results
Is this something that has been considered ? Or could be a great idea ?

My major problem is from my open source package: https://laravel-rest-api.lomkit.com/resources/relationships#morph-to

The workarrond i'm giving:
```php
// You can't declare this relation in Laravel Rest Api
public function commentable(): MorphTo
{
    return $this->morphTo();
}

public function post(): MorphTo
{
    return $this->commentable()->whereHas('comments', function (Builder $query) {
        $query->where('comments.commentable_type', Post::class);
    });
}

public function video(): MorphTo
{
    return $this->commentable()->whereHas('comments', function (Builder $query) {
        $query->where('comments.commentable_type', Video::class);
    });
}
```

This is pretty bad for performances ... But this works

**Comment by GautierDele** [comment]
Nothing new here ?

**Reply by pindab0ter** [reply]
If there is nothing new here, then how should you specify what model you want from a MorphTo relationship, without Laravel having to perform an extra SQL query as it would in the given example?

Because the given example causes two queries to be fired. One to get the `comments` row(s) and one to get the `commentable`s, with an extra `where` referring back to the `comments` table.

The ideal solution would do a check after that first query, determine that the `commentable_type` field isn't what we expected and then just return a relation with a `null` result.

**Comment by murilolivorato** [comment]
I did a article about that  -
https://medium.com/@murilolivorato/mastering-polymorphic-relationships-in-laravel-a-comprehensive-guide-ff3bc3ef2b64
and the source code on git hub -
https://github.com/murilolivorato/laravel-polymorphs-project

**Reply by GautierDele** [reply]
Please read back my discussion, you are out of context

**Comment by pindab0ter** [comment]
I found a solution and wrote a blog post about it:

https://hansvl.nl/blog/constraining-laravels-morphto-relation/

**Reply by MizouziE** [reply]
Seems like a nice approach. Could I convince you to release it as a package and/or PR it?

**Reply by pindab0ter** [reply]
https://packagist.org/packages/pindab0ter/constrained-morph-to-for-laravel

**Reply by MizouziE** [reply]
Awesome! I'll check this out because I have a real world need for this. Feedback pending, but regardless, thank you very much for this man! 🔥 ♥️

**Comment by shaedrich** [comment]
> // You can't declare this relation in Laravel Rest Api

Why?
